tools/autograd/templates/Functions.cpp

```diff
@@ -1970,68 +1970,125 @@ Tensor symeig_backward(const std::vector<torch::autograd::Variable> &grads, cons
+    } else {
+      // Q is ... x M x N, Q.T is ... x N x M and grad_Q is ... x M x N
+      Q_term = at::zeros_like(R, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+    }
+
+    // We want to compute: (rhs_solve_1 . R^{-T})
+    // Note that (rhs_solve_1 . R^{-T}) = (R^{-1} . rhs_solve_1^{T})^{T}
+    // Since R is upper triangular, we can do this using
+    // triangular_solve(rhs_solve_1^{T}, R)^{T}
+    auto rhs_solve_1 = R_term - R_term.transpose(-2, -1) + Q_term - Q_term.transpose(-2, -1);
+    rhs_solve_1 = at::tril(rhs_solve_1, /*k=*/-1);
+    Tensor solve_soln_1;
+    std::tie(solve_soln_1, std::ignore) = at::triangular_solve(rhs_solve_1.transpose(-2, -1), R,
                                                              /*upper=*/true, /*transpose=*/false,
                                                              /*unitriangular=*/false);
-  Tensor grad_A;
-  if (grad_R.defined()) {
-    grad_A = at::matmul(Q, solve_soln_1.transpose(-2, -1) + grad_R);
-  } else {
-    grad_A = at::matmul(Q, solve_soln_1.transpose(-2, -1));
-  }
-
-  // Successive computations involve computation of QQ^{T} which is identity when A is square
-  if (self.size(-1) != self.size(-2)) {
-    Tensor rhs_solve_2;
-    // We use the same trick from above for this computation
-    if (grad_Q.defined()) {
-      rhs_solve_2 = grad_Q - at::matmul(Q, Q_term);
+    Tensor grad_A;
+    if (grad_R.defined()) {
+      grad_A = at::matmul(Q, solve_soln_1.transpose(-2, -1) + grad_R);
     } else {
-      rhs_solve_2 = -at::matmul(Q, Q_term);
+      grad_A = at::matmul(Q, solve_soln_1.transpose(-2, -1));
     }
-    Tensor solve_soln_2;
-    std::tie(solve_soln_2, std::ignore) = at::triangular_solve(rhs_solve_2.transpose(-2, -1), R,
+
+    // Successive computations involve computation of QQ^{T} which is identity when A is square
+    if (A.size(-1) != A.size(-2)) {
+      Tensor rhs_solve_2;
+      // We use the same trick from above for this computation
+      if (grad_Q.defined()) {
+        rhs_solve_2 = grad_Q - at::matmul(Q, Q_term);
+      } else {
+        rhs_solve_2 = -at::matmul(Q, Q_term);
+      }
+      Tensor solve_soln_2;
+      std::tie(solve_soln_2, std::ignore) = at::triangular_solve(rhs_solve_2.transpose(-2, -1), R,
                                                                /*upper=*/true, /*transpose=*/false,
                                                                /*unitriangular=*/false);
-    grad_A.add_(solve_soln_2.transpose(-2, -1));
+      grad_A.add_(solve_soln_2.transpose(-2, -1));
+    }
+    return grad_A;
+  };
+
+  //NOTE: grad of complete QR cannot exist b/c inverse not exist/unique
+  TORCH_CHECK(((r.size(-2) <= r.size(-1) && (!some)) || some),
+		     "The derivative is not implemented when nrows > ncols and complete QR. ");
+
+  auto grad_Q = grads[0];
+  auto grad_R = grads[1];
+
+  auto m = self.size(-2);
+  auto n = self.size(-1);
+
+ if (m >= n) {
+    return square_deep_case_backward(grad_Q, grad_R, self, q, r);
```

Comment:
so what I mean by exactly the same is that when you call `square_deep_case_backward` in this block is the same result as previous implementation.